osbuild/util/pe32p.py

```diff
@@ -0,0 +1,171 @@
+
+
+OptionalFormat = "HBBIIIIIQIIHHHHHHIIIIHHQQQQII"
+OptionalHeader = namedtuple(
+    "OptionalHeader",
+    # Standard fields
+    "Magic "
+    "MajorLinkerVersion "
+    "MinorLinkerVersion "
+    "SizeOfCode "
+    "SizeOfInitializedData "
+    "SizeOfUninitializedData "
+    "AddressOfEntryPoint "
+    "BaseOfCode "
+    # Windows-Specific fields (PE32+)
+    "ImageBase "
+    "SectionAlignment "
+    "FileAlignment "
+    "MajorOperatingSystemVersion "
+    "MinorOperatingSystemVersion "
+    "MajorImageVersion "
+    "MinorImageVersion "
+    "MajorSubsystemVersion "
+    "MinorSubsystemVersion "
+    "Reserved1 "
+    "SizeOfImage "
+    "SizeOfHeaders "
+    "CheckSum "
+    "Subsystem "
+    "DllCharacteristics "
+    "SizeOfStackReserve "
+    "SizeOfStackCommit "
+    "SizeOfHeapReserve "
+    "SizeOfHeapCommit "
+    "LoaderFlags "
+    "NumberOfRvaAndSizes "
+)
+
+
+def seek_to_coff(f, check: bool = False):
+    if check:
+        f.seek(0, io.SEEK_SET)
+        magic = f.read(2)
+        assert magic == b"MZ", "Not a PE32+ file (missing DOS stub)"
+
+    f.seek(0x3c, io.SEEK_SET)
+    buf = f.read(struct.calcsize("I"))
+    (s, ) = struct.unpack_from("I", buf)
+    f.seek(int(s), io.SEEK_SET)
+
+
+def read_coff_header(f, check: bool = True) -> CoffHeader:
+    seek_to_coff(f, check=check)
+    buf = f.read(struct.calcsize(CoffFormat))
+    coff = CoffHeader._make(struct.unpack_from(CoffFormat, buf))
+    assert coff.Signature == b"PE\0\0", "Not a PE32+ file (missing PE header)"
+    return coff
+
+
+def read_optional_header(f, coff: Optional[CoffHeader] = None) -> OptionalHeader:
+    if coff is None:
+        coff = read_coff_header(f)
+
+    buf = f.read(coff.SizeOfOptionalHeader)
+    sz = struct.calcsize(OptionalFormat)
+    assert len(buf) >= sz, "Optional header too small"
+    opt = OptionalHeader._make(struct.unpack_from(OptionalFormat, buf))
+    assert opt.Magic == 0x20B, f"Not a PE32+ file (magic: {opt.Magic:X})"
+    return opt
+
+
+def iter_sections(f, coff: Optional[CoffHeader] = None) -> Iterator[SectionHeader]:
+    if coff is None:
+        coff = read_coff_header(f)
+        f.seek(coff.SizeOfOptionalHeader, io.SEEK_CUR)
+
+    for _ in range(coff.NumberOfSections):
+        buf = f.read(struct.calcsize(SectionFormat))
+        yield SectionHeader._make(struct.unpack_from(SectionFormat, buf))
+
+
+def read_sections(f, coff: Optional[CoffHeader] = None) -> List[SectionHeader]:
+    return list(iter_sections(f, coff))
+
+
+def main():
+
+    if len(sys.argv) != 2:
+        print(f"usage: {sys.argv[0]} FILE")
+        sys.exit(1)
+
+    with open(sys.argv[1], "rb") as f:
+        coff = read_coff_header(f)
+        opt = read_optional_header(f, coff)
+        sections = read_sections(f, coff)
+
+    print(coff)
+    print(opt)
+    for s in sections:
+        print(s)
+
+    last = sections[-1]
+    print(f"{last.VirtualAddress: X}, {last.VirtualSize:X}")
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
Is that intentionally left in? I don't mind it too much, but we don't do that anywhere else, do we?

Reply:
I have done it in:
- [util/lvm2.py](https://github.com/osbuild/osbuild/blob/70bed2e9847328b6faa80b9ddeaea2a6e2758377/osbuild/util/lvm2.py#L618)
- [test/initrd.py](https://github.com/osbuild/osbuild/blob/70bed2e9847328b6faa80b9ddeaea2a6e2758377/test/initrd.py#L181)

It is sometimes quite useful, but if you strongly object, I can take it out.

Reply:
Nah, leave it. I don't mind.